### PR TITLE
[Slim] Add Basic authentication middleware

### PR DESCRIPTION
**Pull request by ybelenko**
### PR checklist

- [x] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [x] Ran the shell script under `./bin/` to update Petstore sample so that CIs can verify the change. (For instance, only need to run `./bin/{LANG}-petstore.sh` and `./bin/security/{LANG}-petstore.sh` if updating the {LANG} (e.g. php, ruby, python, etc) code generator or {LANG} client's mustache templates). Windows batch files can be found in `.\bin\windows\`.
- [x] Filed the PR against the [correct branch](https://github.com/OpenAPITools/openapi-generator/wiki/Git-Branches): `master`, `3.1.x`, `4.0.x`. Default: `master`.
- [x] Copied the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) to review the pull request if your PR is targeting a particular programming language.

### Description of the PR
Basic auth middleware verifies token from `Authorization` header.
If token not verified it returns `401` http status code and quits script execution.
~User needs to implement `verifyCredentials` method in `AuthBasic.php`. By default this method always returns `false` to avoid security holes on production.~ My code swapped with [tuupola/slim-basic-auth](https://github.com/tuupola/slim-basic-auth) package.
For a quick start, user needs to implement own `authenticator` in `SlimRouter` class:
```php
$basicAuth = new HttpBasicAuthentication([
    "secure" => false,
    "authenticator" => function ($arguments) {
        $user = $arguments["user"];
        $password = $arguments["password"];
        // write your own code here, compare credentials with database etc.
        return false;
    }
]);
```

Test endpoint with Basic auth:
```shell
$ curl -X POST "http://petstore.swagger.io:80/v2/fake" \
    -H "accept: application/json" \
    -H "Content-Type: application/x-www-form-urlencoded" \
    -d "number=33&double=68.9&pattern_without_delimiter=FooBar&byte=fdfdfdf" -i
```

Output:
```shell
HTTP/1.1 401 Unauthorized
Date: Tue, 24 Jul 2018 08:54:42 GMT
Server: Apache/2.4.23 (Win64) PHP/7.2.7
X-Powered-By: PHP/7.2.7
WWW-Authenticate: Basic realm="Protected"
Content-Length: 0
```

cc @jebentier @dkarlovi @mandrean @jfastnacht @ackintosh

### CODE REVIEW

**Comment by ybelenko** [comment]
At the last moment I've found [slim-basic-auth](https://github.com/tuupola/slim-basic-auth) package.
Obviously this package more stable than my script, but I've noticed few disadvantages:
1. Big dependency list
2. Hardcoded `401` response when credentials doesn't match
3. Too many settings
4. Plugin allows unsecure connection for only `relaxed` hosts
5. We need two more middlewares(AuthApiKey, AuthOAuth) and I don't think that existen plugins will follow same code patterns as described one.

**Comment by ybelenko** [comment]
@wing328 @ackintosh One more thing. If we use 3rd party package then user have only settings without access to middleware source files. I'm still thinking what approach I should use.

**Comment by ybelenko** [comment]
I can remove revert commits and wipe off custom implementation from history if needed.

File: samples/server/petstore/php-slim/phpunit.xml.dist [hunk 1]
```diff
@@ -19,8 +19,8 @@
     </testsuites>
     <filter>
         <whitelist processUncoveredFilesFromWhitelist="true">
-            <directory suffix=".php">./lib/Api</directory>
-            <directory suffix=".php">./lib/Model</directory>
+            <directory suffix=".php">./lib//Api</directory>
+            <directory suffix=".php">./lib//Model</directory>
```
Comment (wing328) [inline samples/server/petstore/php-slim/phpunit.xml.dist]: @ybelenko I believe the double forward slash won't cause any issue. Please correct me if I'm wrong.

Comment (ybelenko) [reply samples/server/petstore/php-slim/phpunit.xml.dist]: Hmmm, and what are benefits from double forward slash, may I ask? :smile:

Comment (wing328) [reply samples/server/petstore/php-slim/phpunit.xml.dist]: Well I do not know. I spotted the double forward slash and just asking to make sure phpunit still works fine with it.

Comment (ybelenko) [reply samples/server/petstore/php-slim/phpunit.xml.dist]: Just checked on Win8.1, phpunit works fine. These slash changes is result of #610  . I just use `{{apiSrcPath}}` and `{{modelSrcPath}}` which defined in AbstractPhpCodegen.

Comment (wing328) [reply samples/server/petstore/php-slim/phpunit.xml.dist]: 👌

**Comment by jmini** [comment]
This PR breaks the Shippable CI build.

**Comment by ybelenko** [comment]
@jmini Sad to hear it. Any ideas what's wrong and how to fix it?

**Comment by jmini** [comment]
@ybelenko thank you for asking, no problem.

I have run `bin/utils/ensure-up-to-date` and pushed corresponding commit: https://github.com/OpenAPITools/openapi-generator/commit/d8ea28e147ac84a14f1765c4a7f9fabc32737d8b

It is now OK on `master`.

**Pull request merged**